### DISCUSSION TITLE
New user - have a few questions about hardware setup

### DISCUSSION THREAD

**Discussion by thecourteener** [awawa-dev/HyperHDR#486] [Q&A]
Hi guys, sorry to trouble you! I'm just about to start my HyperHDR journey, have purchased most of the parts but just have a few questions about the hardware setup. Looks an amazing product btw, absolute credit to you awawa-dev! So these are the parts I have purchased and the setup I will have;

Ezcoo Splitter - Rullz MS2109 USB 2.0 Grabber - Raspberry Pi 3B+ (Powered by it's own official charger) - ESP32 module with CP2102 chip - SK6812 5m LED Strip (Powered by it's own Meanwell 5v/18amp/90w power supply)

Does that all look ok? So the questions I have;

1. I've also purchased a logic level shifter as advised on here. Will this go in-between the ESP32 and the LED strip or in-between the Raspberry Pi and the ESP32? 
2. I plan to use HyperSPI and connect the ESP32 to the Raspberry Pi using 4 SPI cables, because I assume from what i've read, this will give me the fastest connection? If I do this, can I also connect the ESP32 to the Raspberry Pi via USB as the same time to provide power and ground to it? If I connect the SPI cables and USB at the same time, how does HyperSPI know which to use for communication?
3. Will I need a ground cable directly between my Raspberry Pi and my LED strip, or will the fact the ESP32 is driving the strip mean I don't need this?

Many thanks in advance and sorry for all the questions!

**Comment by awawa-dev** [comment]
Hi

> I've also purchased a logic level shifter as advised on here. Will this go in-between the ESP32 and the LED strip or in-between the Raspberry Pi and the ESP32?

Always use it for LED signal that comes from the LED strip driver to the LED strip so in your case, it should be placed between ESP32 and the LED strip.

> I plan to use HyperSPI and connect the ESP32 to the Raspberry Pi using 4 SPI cables, because I assume from what i've read, this will give me the fastest connection? If I do this, can I also connect the ESP32 to the Raspberry Pi via USB as the same time to provide power and ground to it? If I connect the SPI cables and USB at the same time, how does HyperSPI know which to use for communication?

In term of LED refresh rate there is little advantage of HyperSPI over HyperSerialESP32. Of course the latency is lower thanks to SPI speed but the main benefit is we don't use USB which can affect other USB2.0 devices on Rpi. The problem is described on HyperSPI homepage. HyperSerialESP32 is easier to setup but for you need to use custom 1Mb speed for CP2102: this chip is not on the recommended boards on HyperSerialESP32 page but it should rather work.

> Will I need a ground cable directly between my Raspberry Pi and my LED strip, or will the fact the ESP32 is driving the strip mean I don't need this

You need a ground cable between LED strip (if it has independent power supply) and your LED driver so in your case ESP32.

**Comment by thecourteener** [comment]
Many thanks for getting back to me awawa-dev, really appreciate that! In regards to the ESP32 board I have purchased then, would you return the one i've purchased with the CP2102 chip and get one with a CH9102x or CP2104 chip instead and run it off USB? Or should I just stick with the one i've got with CP2102 and use HyperSPI? If I stick with the one i've got, how am I best powering the ESP32 board?

**Comment by awawa-dev** [comment]
It should be fine for HyperSPI. But if we can have a more universal board of better performance, it should be choose instead because now you are limited to HyperSPI only. You can power ESP from Rpi connecting it to the Rpi USB port.

**Comment by thecourteener** [comment]
amazing thanks so much for your help mate :) I think i'll send it back and get another board then!

**Comment by awawa-dev** [comment]
Be careful choosing boards... almost all with proper specs will be fine but sometimes they can have bad design as you can read here: https://github.com/awawa-dev/HyperHDR/discussions/401#discussioncomment-4653269 I tested and recommended two ESP32 & ESP32-S2 boards, [you can find them on HyperSerialESP32 page](https://github.com/awawa-dev/HyperSerialESP32#example-of-supported-boards) ESP32 MH-ET Live is old and stable solution, ESP32-S2 mini is cheap & new but very promising board (time will tell).

**Comment by thecourteener** [comment]
thanks so much awawa! Just out of curiosity so I can get faster shipping, do you think this board would be ok? It has CH340 but i'm not sure if there's a difference between CH340 and CH340G

https://www.az-delivery.uk/en/products/esp32-dev-kitc-v2-ai-thinker-soldered

**Reply by awawa-dev** [reply]
Not sure: it's ESP from Ai-Thinker not Espressif if that make any difference and its CH340 has unknown specification.

**Comment by thecourteener** [comment]
ok mate i'll go with the one you suggested then, can't go wrong then :) thanks again

**Comment by thecourteener** [comment]
Awawa can I ask for some advise please mate. I got all my hardware and set everything up today, can’t get the led lights to work at all. Here’s my setup;

Ezcoo splitter
2109 usb grabber
SK6812 neutral LED’s (264 led’s)
Raspberry Pi 4 
Lolin S2 mini 
Logic level shifter as advised 
5v 18amp 90w power supply 

I’m feeding power at the start of the strip, the middle, and the end. 

I flashed the lolin s2 mini with hyperserialesp32 using esptool.py. Video capture is working absolutely fine in hyper hdr but absolutely nothing happens with the led’s. I did mess around with controller type in the LED hardware settings and managed to get one LED to come on white in the middle of the whole thing which was really odd. 
what should the controller type be set to if I’m powering and connecting the s2 mini via usb from the raspberry pi?

**Reply by thecourteener** [reply]
Just noticed in the logs on Hyper HDR it keeps saying this, is this my issue? This is with the controller set to adalight

2023-02-06T22:18:49.021Z [LEDDEVICE_ADALIGHT] Could not detect HyperSerialEsp8266/HyperSerialESP32 device

**Reply by awawa-dev** [reply]
Something is not right with ESP32-S2 in first place and that single lit LED is probably just ghost light. Make sure you have selected proper COM and please: FULL logs not just one line. Post also your full command you used to flash firmware to ESP32-S2 with `esptool.py`. Did you try to reset ESP?

**Reply by thecourteener** [reply]
thanks for replying so quick awawa! To flash the S2 I put it into device firmware upgrade mode first then typed;
esptool.py --port COM4 erase_flash
I then typed;
esptool.py --port COM4 --baud 1000000 write_flash -z 0x1000 firmware_esp32_s2_mini_SK6812_RGBW_NEUTRAL.bin

does that look right?

**Reply by awawa-dev** [reply]
No, you missed one critical zero and probably board is locked in bootloader mode now:
`esptool.py write_flash 0x10000 firmware_esp32_s2_mini_SK6812_RGBW_NEUTRAL.bin` (four zeroes)

> esptool.py --port COM4 erase_flash

that is usually not neccesery

**Reply by thecourteener** [reply]
ahhh damn, is there any way to take it out of bootloader mode?

**Reply by awawa-dev** [reply]
Just re-flash it using the command I gave you. You may need to press `reset` + `en` buttons, then release `reset`, next release `en` and start flashing it with esptool.py. Remember to reset it afterwards: esptool.py wont be able to do it.

**Reply by awawa-dev** [reply]
If wont work  (HyperHDR wont detect it again) then use that command which usually is not necessary
> esptool.py --port COM4 erase_flash

and repeat reflashing procedure. Remember to set proper speed in HyperHDR (the same you compiled the firmware for)

**Reply by thecourteener** [reply]
ok thanks mate i'll give that a try. Just one last question, i'm new to all this sorry......is that all I need to flash on the S2 mini or do I need to flash a bootloader as well?

**Reply by awawa-dev** [reply]
not sure.. never missed when uploading the firmware in the correct location in flash ;) Try it first.

**Reply by thecourteener** [reply]
haha i'll give it a try :) thanks mate, i'll let you know

**Reply by thecourteener** [reply]
managed to re-flash the S2-mini correctly mate, hyperhdr see's it now but nothing is working properly. Only the led's down the left side of the tv and the top of the tv come on and even then they basically just show white or just really really pale colours.
I think i've reached my limits on it because not a clue what to do or what to check. Thanks for all the help you've given me, i'll resign to another project i've started and failed haha

**Reply by LaytexLlama** [reply]
It sounds like you're in the exact same boat as me and a couple other people's discussions I've read. Its an amazing product and project but seems a little complex. I'm going to give mine another crack soon when the new ESP boards come in and hope for the best haha.. Pretty much just reading every discussion and thread I can find hoping I can figure it out and if I can, try compile it all in 1 place.

**Reply by thecourteener** [reply]
yessssssssssssssssssss it's a miracle, I have it all working :) :) :) :) took the tv off the wall last night and noticed the data link on one of the corners where i'd chopped the LED strip to put it along the top and side had ripped the pad off on the led strip. I must have been a bit too heavy handed pushing the cables to make them neat. Cut a couple of LED's out the strip and replaced them, soldered it all back together and it's all working perfect now :) 

@awawa-dev I need your help again now mate haha. So at first the colours looked really washed out, i've fixed this by increasing the contrast on the video capture settings page and things look good now. I didn't realise I could do this at first so I changed a few of the values in those colours boxes that have (255,255.0) in them etc. Is there any way I can put them back to default because I stupidly didn't make a note of what they were originally? Is there any other settings I should be changing to make things perfect? I'm using one of the newer usb 3.0 MS2130 capture devices running into a Raspberry Pi 4. I've currently set it use YUV 640x480 20fps, is that ok or should I increase this? Is there any other settings your recommend changing?

**Reply by awawa-dev** [reply]
>  So at first the colours looked really washed out

@thecourteener that usually happens (but not limited to) when you capture HDR signal and you have 'HDR tone mapping' disabled in HyperHDR e.g. [before and after](https://camo.githubusercontent.com/c50df292fcb315a89b4a518385f77491159cace25a73b4354b8cf82aae87421d/68747470733a2f2f692e706f7374696d672e63632f5673625a724742782f6366696e616c2e6a7067) ..must check what do you have in the live preview. It's explained here: [Configuring video & system grabber section](https://github.com/awawa-dev/HyperHDR/wiki/Quick-start#configuring-video--system-grabber) MS2130 doesnt have yet dedicated LUT table but you can try MS2109 LUT just remember to set brightness/contrast/saturation as in the manual.
https://www.hyperhdr.eu/2022/04/usb-grabbers-hdr-to-sdr-quality-test.html#chapter4

To reset all your HyperHDR settings to default just delete `.hyperhdr` directory in your home folder e.g. (/home/pi/.hyperhdr). May be hidden for some file explorers.

**Reply by awawa-dev** [reply]
> YUV 640x480 20fps

resolution is fine but you need higher framerate if possible

**Reply by thecourteener** [reply]
thanks so much!! :) i'll give all that a try

**Reply by awawa-dev** [reply]
@thecourteener I've just receive MS2130. Beta version of dedicated LUT table is already available here: https://github.com/awawa-dev/HyperHDR/discussions/499 Don't change brightness, contrast or saturation for this grabber!

**Reply by thecourteener** [reply]
@awawa-dev ahhh amazing, I had set things up with the MS2109 LUT table and manually set the brightness/contrast/saturation like you suggested and it's been looking great but i'll try the beta dedicated LUT table later :)
Just been reading your updated ultimate guide, which is amazing! you must have put a lot of work into that. I haven't touched the smoothing options on my setup yet, I just left it as default so I will go through the update frequency config you have described later :)
Just one last question mate......the firmware I flashed on my ESP32-S2 lolin mini........is it already configured to use the 5mb serial port speed it is capable of or do I need to re-compile the firmware to make use of it?

**Reply by awawa-dev** [reply]
@thecourteener thanks! To have 5Mb speed you need to recompile the firmware: I don't want to confuse users preparing many various versions of firmware in the release section. And 2Mb allows to fully saturate Neopixel data line and to reach maximum refresh rate. Still with 5Mb you can have lower latency:
```
for 2Mb speed 400 RGBW LEDS => 400*3 bytes => 400*3*8 bits => 9600 / 2000000 => additional 5ms transport latency
for 5Mb speed 400 RGBW LEDS => 400*3 bytes => 400*3*8 bits => 9600 / 5000000 => additional 2ms transport latency 
```
update: forget that over serial it's always rgb not rgbw

**Reply by awawa-dev** [reply]
> is it already configured to use the 5mb serial port 

@thecourteener 
To clarify it : that's *undocumented* feature :) It works also for me but have completely no idea why because the port in the standard firmware is initialized using 2Mb...but CDC in ESP32-S2 does what it wants anyway and adjust the speed also for 5Mb. Better to be safe and prepare dedicated firmware for this speed.

**Reply by thecourteener** [reply]
brill thanks awawa, you da man :) How do I recompile the specific firmware 'firmware_esp32_s2_mini_SK6812_RGBW_NEUTRAL.bin'? I've recompiled before using your tutorial and making changes to platformio.ini but how do I recompiles the s2_mini firmware you've already done?

**Reply by awawa-dev** [reply]
OK, I see that you have your fork here: https://github.com/thecourteener/HyperSerialESP32 It's two commits behind the upstream but these commits have only modified readme.md to provide extended manual and don't bring any changes to the firmware. You have already switched it to 5Mb speed (https://github.com/thecourteener/HyperSerialESP32/commit/a1347704ac5e54e879eaf0e7e207fd0122f6d718) and all firmware in this package should work at 5Mb (including esp32-s2) https://github.com/thecourteener/HyperSerialESP32/actions/runs/4106334991 If you flashed it then you can use 5Mb.

**Reply by thecourteener** [reply]
bloody amazing!! if you lived nearby i'd buy you a few pints haha

**Comment by thecourteener** [comment]
@awawa-dev mate i'm so sorry to keep mithering you, just struggling to get my head around this recompiling. To change the serial port speed, If all you need to do to recompile is change the DSERIALCOM_SPEED= in the source code, then why is there different firmware's for different models of ESP board and LED's? Do you need to delete all the other models/led's in the source code as well as changing DSERIALCOM_SPEED= to make it specific to your device? I'm just confused as to what is different in all the different firmware bin files because they all have different kb sizes? If I recompile the main source code, surely it won't be specific to my model (s2 mini). Sorry I know i'm probably sounding dumb

**Reply by awawa-dev** [reply]
First : each LED type requires special class to support it from my fork of Neopixelbus library.  Second: esp32 and esp32-s2 are quite different architecture. So 5 LED strip types * 2 types of ESP = 10 completely different firmwares. If you change the speed (in one place, it's global define) then you change the speed for all of them. You don't need to delete other models, just use the one for your type of LED.

**Reply by awawa-dev** [reply]
> if I recompile the main source code, surely it won't be specific to my model (s2 mini)

Hope you extract the archive zip before flashing...then you will find the correct .bin firmware by the name.

**Reply by thecourteener** [reply]
god @awawa-dev please ignore me, i'm being so dumb!!!! I didn't realise that when you download the artifact it comes down as a zip file and contains all the different firmwares haha. This is where i've been getting confused, I thought when you downloaded the artifact it contained 1 single firmware bin file. That's why I couldn't understand what made it specific to my model. D'oh so sorry!! I've got it now haha

**Reply by awawa-dev** [reply]
this is not the first such mistake ;) I have already clarified it on the blog... I still need to update the wiki

**Comment by thecourteener** [comment]
@awawa-dev i've had a really weird issue mate since i've installed version 19 instead of the 19 beta. Normally when I turn the tv off, because it powers my firestick by usb, hyperhdr loses it's video signal so the led's turn off fully. Now when I turn the tv off, the led's all turn a dim white and stay like that. Any ideas? I've re-flashed my s2 mini and reinstalled hyperhdr twice but it's made no difference. Was fine before

**Reply by awawa-dev** [reply]
turn on/off LED in the remote tab works correctly?

**Reply by thecourteener** [reply]
yeah that works fine, they go off fully. Does that rule out hyperhdr then?

**Reply by awawa-dev** [reply]
Probably you need to adjust your settings and disable (set zero) Backlight threshold in processing (but it's not recommended if you want the  anti-flickering filter in the Smoothing to work at its best)

**Reply by thecourteener** [reply]
@awawa-dev it's really really weird this issue of the lights staying on at all times. If I turn off the LED in the remote tab they turn off properly like they should and the lights fully go off. If I disable video capture, they turn off properly like they should and the lights fully go off. If I turn off the input source though, the lights stop displaying the colours etc but they all stay on a dim white. The weirdest thing is even if I fully disconnect the LED's from the ESP32 so they're getting no data signal whatsoever, they still stay on a dim white. Do you think it's just current from the power supply causing it?

**Reply by awawa-dev** [reply]
No it's just your software configuration as I explained. Please disable minimal backlit in the HyperHDR processing tab (which prevents the LEDs to render black color which effects in turning off LED, but the antiflickering filter may work not at 100% effecient then) or provide full HyperHDR logs taken after 2-3 minutes after restart (paste it on pastebin.com).

**Reply by thecourteener** [reply]
@awawa-dev sincere apologies for not listening to you, you were right. The backlight threshold was set to 2% on mine. I've put it to 0% and the lights turn off absolutely fine now :) Will this affect the smoothing now i've set it to 0?

**Reply by awawa-dev** [reply]
If you video source has tendency to provide some noice/dithering and you are using high-refresh rate LEDs then you may expect blinking when the colors are near 0, because anti-flickering filter if it's enabled does not work when the target color is 0. So in scale 0-255 RGB, when there is a noise around very dark gray at 1, the LED will be blinking 0 (turn off) - 1 (turn on, quite dramatic difference) and so on. That's why even minimal backlight is needed to disable such effect.

**Reply by thecourteener** [reply]
ok thanks mate, yet again appreciate all your help massively :)

**Comment by Jesper80** [comment]
Hi Guys. I have been looking at HyperHDR and it looks really good. After reading through the very extensive forum post "Ultimate Guide 2023" I still have a few questions.

I will be using my Windows PC for all the streaming, mostly youtube and Plex. Do I then need a USB capture card (MS2130), or can I use the DirectX 11 screen capture feature instead with HyperHDR running on the Windows PC? Or is both options available, just that the capture card is the preferred option?

I would prefer not to put a capture card in between my HDMI 2.1 signal and the TV. Could I just put the MS2130 capture card into a second HDMI port and some how mirror the first screen in Windows to also output to the capture card (maybe at a lower resolution?)

What do I need the ESP32 device for? Some youtube videos that I've seen is just a RPI connected directly to the LED:s

I have an unused RPI 4 laying here. Do I even need that if I'm using a Windows computer (with Screen capture or capture card) and an ESP32. Or is it the other way around that I can use the computer and just a RPI?

My TV is a 65" LG OLED. Is there any benefit in using 144 LED/m  (662 LED:s) vs 60 LED/m (276 LED:s)?

**Reply by awawa-dev** [reply]
> I will be using my Windows PC for all the streaming, mostly youtube and Plex. Do I then need a USB capture card (MS2130), or can I use the DirectX 11 screen capture feature instead with HyperHDR running on the Windows PC? Or is both options available, just that the capture card is the preferred option?

Exactly. Both options are available, but the grabber is preferred. There's no guarantee that software capturing won't cause issues/shuttering in some apps, but it usually works fine for our users. At least no one reported the problems for quite some time.

> I would prefer not to put a capture card in between my HDMI 2.1 signal and the TV. Could I just put the MS2130 capture card into a second HDMI port and some how mirror the first screen in Windows to also output to the capture card (maybe at a lower resolution?)

MS2130 won't allow you to turn on HDR in Windows. You need a HDMI splitter to "fool" the video source that MS2130 is capable to receive HDR signal (wash-out colors but HyperHDR can fixed it), even if it won't split the signal with this device. Or use some other EDID emulator.

> What do I need the ESP32 device for? Some youtube videos that I've seen is just a RPI connected directly to the LED:s

Well, that's bad solution (3.3V logic connected directly to 5V LED logic, but it may work for some systems...matter of luck). Use the ESP32 or ESP32-S2 mini board from latest tutorial on my blog and the level shifter if you want the solution that fulfill LED tech specs.

> I have an unused RPI 4 laying here. Do I even need that if I'm using a Windows computer (with Screen capture or capture card) and an ESP32. Or is it the other way around that I can use the computer and just a RPI?

HyperHDR can run on Windows using USB output to ESP32 with HyperSerialESP32. You can test this solution first, without Rpi.

> My TV is a 65" LG OLED. Is there any benefit in using 144 LED/m (662 LED:s) vs 60 LED/m (276 LED:s)?

144 should provide better saturation and details but much brighter overall effect (it may be too bright). You can think about the compromise: 96LEDs/meter. Didn't test 144 LEDs/meter setup yet but I'm using 96LEDs/meter in my home solution.

**Reply by Jesper80** [reply]
Thank you so much for your extensive reply.
 
> MS2130 won't allow you to turn on HDR in Windows. You need a HDMI splitter to "fool" the video source that MS2130 is capable to receive HDR signal (wash-out colors but HyperHDR can fixed it), even if it won't split the signal with this device. 

Okey, I see. So in that case I would need both a HDMI splitter and an MS2130. But does this actually have to be sending a signal to the TV, or could I use that setup on a separate HDMI port, somehow mirrored in windows, maybe at a lower resolution?

>Or use some other EDID emulator.

Do you mean that there are other emulators than the MS2130 that would support HDR, and therefore not require a separate splitter in between?
 
> Well, that's bad solution (3.3V logic connected directly to 5V LED logic, but it may work for some systems...matter of luck). Use the ESP32 or ESP32-S2 mini board from latest tutorial on my blog and the level shifter if you want the solution that fulfill LED tech specs.

Alright. Yes that makes sense. And If I'm going to start off with just the screen capture anyway without the RPI, I might as well get me an ESP32 board.

I see that there is a newer version available as well. ESP32-S3. Supposed to be dual core and some more SRAM and ROM. Would this provide any benefit or future proofing? Is it even supported?

> HyperHDR can run on Windows using USB output to ESP32 with HyperSerialESP32. You can test this solution first, without Rpi.

Will do! Just for me to make sure that I fully understood everything. This is the setup I imagine.

* HyperHDR running on the windows computer with screen capture and HyperSerialESP32.
* USB3 cable from the PC to an ESP32-S2(3) board.
* LED stip connected to the ESP according to your blog.

Or do I miss anything?
 
> 144 should provide better saturation and details but much brighter overall effect (it may be too bright). You can think about the compromise: 96LEDs/meter. Didn't test 144 LEDs/meter setup yet but I'm using 96LEDs/meter in my home solution.

96 sounds like a really good idea. But in case they are too bright, can HyperHDR dim them somehow?

The LED:s I was originally looking at was the SK6812 from btf-lighting. But they don't seem to offer a 96 LED version. Do you have any brand/version of this strip that you could recommend that is available on amazon or aliexpress?

Do you inject voltage at a certain interval on the strip to prevent voltage drop / discoloration?

**Reply by awawa-dev** [reply]
> Do you mean that there are other emulators than the MS2130 that would support HDR, and therefore not require a separate splitter in between?

I didnt test any such device. MS2130 is simply HDMI1.4 grabber so it wont report HDR capabilties to the video source, and video source will disable it (even if the other mirroring device supports it). Let's just assume you need a HDMI splitter, either when the signal is forwared to the TV or not (first you need to clone EDID from the TV anyway for Feintech, Ezcoo has also switches). Base scenario assumes that HDMI splitter is providing the signal to the TV and the grabber at the same time. If you want HDMI2.0 grabber that reports HDR capabilities (even if it produces washed-out colors) then I recommend Elgato HD60X.

> The LED:s I was originally looking at was the SK6812 from btf-lighting. But they don't seem to offer a 96 LED version. Do you have any brand/version of this strip that you could recommend that is available on amazon or aliexpress?

you have one here: https://aliexpress.com/item/32664925043.html As you probably understand, I can't guarantee the quality because I have nothing to do with the seller and the quality very often changes on Aliexpress even depending on the batch. btf-lighting is the brand just as the others: the only one LED strip that has broke for me almost instantly was came from this manufacturer. Of course no warranty.

> 96 sounds like a really good idea. But in case they are too bright, can HyperHDR dim them somehow?

You can but it's not neutral for the quality. Minimal brightness (for RGB(1,1,1) ) cannot be changed: the more LEDS, the higher brightness.

> Do you inject voltage at a certain interval on the strip to prevent voltage drop / discoloration?

No, only beginning and the end. sk6812 rgbw is very resistant to the voltage drop (more here: https://www.pololu.com/product/2527 , of course for typical application: " The SK6812’s color and brightness are largely independent of its supply voltage from 3.5 V to 5 V, meaning that voltage drops caused by the resistance in long power connections are less likely to have a visible effect.")

**Reply by awawa-dev** [reply]
> I see that there is a newer version available as well. ESP32-S3. Supposed to be dual core and some more SRAM and ROM. Would this provide any benefit or future proofing? Is it even supported?

ESP32-S3 is not supported at least for now. Stick to the recommended boards if you want to avoid traps completely unpredictable for the standard user.

> Or do I miss anything?

3.3v to 5v level shifter?
